### DISCUSSION TITLE
Page tree validation and repair

### DISCUSSION THREAD

**Discussion by m-holger** [qpdf/qpdf-dev#10] [General]
Having looked at QPDF_pages recently I think there are a couple of points worth thinking about:

- getAllPages does a fair bit of checking and repairs. Given
  ```cpp
    // By default, QPDF will try to recover if it finds certain types of errors in PDF files.  If
    // turned off, it will throw an exception on the first such problem it finds without attempting
    // recovery.
    QPDF_DLL
    void setAttemptRecovery(bool);
  ```
  should getAllPages respect `attempt_recovery`?

- getAllPages makes certain repairs, e.g. /TypeName entry, without checking anything else. Would it make sense to check whether the object looks like a page / pages object before attempting to repair.

- some errors are sometimes an exception and sometimes completely ignored, e.g invalid /Count depending on whether flattenPagesTree was called. 

- some errors that seem fixable unconditionally cause exceptions, e.g. loops in the pages tree.

**Comment by m-holger** [comment]
Further thoughts:

It seems pretty clear that invalid pages trees cause a fair bit of trouble in a number of places including `QPDFWriter` and the linearization code. It also seems clear to me that it would be far easier to prevent the problems by inspecting and repairing the pages tree early rather than tracking down the problems caused in far away places. It would therefore appear eminently sensible to call `getAllPages` at the end of `QPDF::parse`.

There is a tiny snag though: the change would be API breaking. Users are required to call `updateAllPagesCache` if they manipulate the pages tree directly after calling `getAllPages`, but not if direct manipulation precedes such call. Having qpdf itself call `getAllPages` earlier than previously has therefore the potential of breaking existing code.

A possible solution to the problem is to create a private API replacement for `getAllPages` which is going to be used by qpdf internally.  `getAllPages` can than be modified to call `updateAllPagesCache` if `!ever_called_get_all_pages`  and than return the result of the private API replacement.

**Comment by m-holger** [comment]
From TODO.md:

> Consider generating a non-flat pages tree before creating output to better handle files with lots
  of pages. If there are more than 256 pages, add a second layer with the second layer nodes having
  no more than 256 nodes and being as evenly sizes as possible. Don't worry about the case of more
  than 65,536 pages. If the top node has more than 256 children, we'll live with it. This is only
  safe if all intermediate page nodes have only /Kids, /Parent, /Type, and /Count.

Non-flat trees was one of the other points that occurred to me, and I did do some extensive research (i.e. I looked at the PDF spec and the qpdf manual), which indicates that 6 - 10 kids may be more typical (based on the extensive sample).

Where the tree is already flat there should be no issues. This includes all cases where qpdf has added or removed nodes,

For other PDF files there is existing code to flatten the tree which will deal with issue of inherited attributes. This leaves of course the question whether qpdf should rebalance the tree, and if so, how much control the user, either at CLI or QPDFWriter level, should have.

**Comment by jberkenbilt** [comment]
I will give you a more comprehensive reply, but I wanted to mention two things right off:
1. I have long considered exposing a reference to the internal pages cache the worst qpdf API mistake I made.
2. A while ago, I started work on completely redoing pages trees with the goal of balancing, etc., but I abandoned the effort. If you want to see where I was going, take a look at refs/attic/QPDFPagesTree in git. (e.g. `git fetch qpdf refs/attic/QPDFPagesTree:qpdf-pages-tree`). I may have left a trail behind in TODO. I can't remember. This was work done in spring 2022.

**Comment by jberkenbilt** [comment]
I'm okay with attemptRecovery being broadened to handle things other than xref recovery. It's a bit slippery as there may be things qpdf does unconditionally already, but I think page tree recovery can be elevated to this.

If you want to add a heuristic to check whether something is a page before changing its type when it is encountered in the pages tree, that's okay, but I'm not sure there's a strong case for it unless you are going to exclude it from the pages tree if you don't change its type. I think we use /Type /Page to stop at page boundaries when copying foreign objects. I don't think we necessarily remove /Type /Page from things we don't encounter in the pages tree. It's hard to know for sure that something's a page. There aren't many required fields. (Maybe just MediaBox? I think you've looked at the spec more recently than I have.) I'd hate to eject a page that's missing a required key if PDF viewers accept it. I've generally taken a policy of not fixing things that qpdf doesn't have to fix to do its job because I don't want qpdf to make the situation worse for broken files with regard to being viewable in other places. In other words, I don't like getting bug reports that say, "I could open this in Acrobat before I ran it through qpdf, and now I can't."

I don't have a blanket opinion about what's an unconditional exception and what's fixable. If it's safe to fix, we should fix it I guess. I think that my original goal of qpdf being a checker has long been overcome by events and its role as a PDF repair tool is vastly more dominant.

Regarding flattening/rebalancing, one of the reasons behind my abandoned work of 2022 was that I was hoping to avoid ever flattening the pages tree. My preference would be for qpdf to leave the pages tree structure alone as much as possible. It would be best if qpdf didn't flatten or rebalance the pages tree unless pages are being added or removed, but there are already exceptions to that. In those cases where qpdf flattens the pages tree now, it would be okay if we created balanced pages trees, though it's a decent amount of trouble. I don't know where the 6-to-10 children comes from. My impulse would have been more like 100 children, but that's not based on anything. I think it's convenient for files with fewer pages to have flat pages trees just because it makes it somewhat easier to work with the files by hand, but this is probably not a big deal.

Anyway, I don't feel too strongly about it, other than that I don't think qpdf should mess with the pages tree if no pages are added or removed and no option that substantially reorganizes the file (like linearization or generation of object streams) is specified. So straight qpdf in.pdf out.pdf or qpdf --qdf in.pdf out.pdf should not touch the pages tree. This is especially important for things like qdf -- we really want the file to look as much as possible like the original from a structural standpoint so we don't take away our ability to look at weird files or inspect the original pages tree. It's also essential that qpdf --json doesn't make any gratuitous changes.

All this is to say that even repair of the pages tree should be triggered only by doing something with the pages tree. If qpdf doesn't have any reason to traverse the pages tree, then it should not touch it or repair any pages because doing so removes the ability to use qpdf as a tool for inspecting broken files and trying to see what's wrong with them, and that does remain an important use case for qpdf.

Hopefully I haven't contradicted myself too much.

**Reply by jberkenbilt** [reply]
If you have other thoughts or questions that require my input, please let me know. I don't know if I've addressed all your concerns or possibly raised other ones.

**Reply by m-holger** [reply]
> My preference would be for qpdf to leave the pages tree structure alone as much as possible. It would be best if qpdf didn't flatten or rebalance the pages tree unless pages are being added or removed, ...

That was my initial view too. I would add to the 'unless` exception that the pages tree needed repair. I'll pick that up in #1

> In those cases where qpdf flattens the pages tree now, it would be okay if we created balanced pages trees, though it's a decent amount of trouble.

I don't think it would be too bad hard as long as rebalancing is done right before writing. (Not sure I would put it up too high on the list of priorities.) Not flattening the tree before insertions/deletions would be tricky and I am pretty sure would have serious performance implications.

>  I don't know where the 6-to-10 children comes from. 

I've extended my already extensive research to a 3rd file. 10 is FrameMaker, 6 is pdfTex. (I was also surprised. I wouldn't have guessed as high as 256, but perhaps 32 or 64).

**Reply by jberkenbilt** [reply]
I've looked at many very large PDF files over many years, and I "feel like" it was more than 10ish but not has high as 100. Alas, I have walked away from a very large source of very large PDF files at my old job....anyway, smaller limits are better for memory-starved systems that load things as needed and discard them, but those systems are less common. I don't think the exact number is important. I'm feeling like 50 is as good as any. It means you have to break 2500 pages before you need a third layer. But, as you say, this is low priority. My pages tree object was trying to make it efficient to do insertions, but it's not really....you're probably right that maintaining as flat internally and balancing right at the end would be the way to go if we wanted to do that. That said, in all the years of maintaining qpdf, I have never had even one bug report or feature request about balanced page trees.
